### PR TITLE
Add slicing and masking for new reduction framwork 

### LINKED ISSUE #19221: Introduce SANS Slicing, Cropping and Masking
This issue addresses the next SANS reduction framework elements:
* event slicing
* cropping 
* masking

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Adds slicing and masking for the new reduction framework

**To test:**

The main thing is a thorough code review.

The entire functionality is already with the scientists for validation. I am not sure how much sense it would make to have functional testing outside of the entire reduction chain.

Fixes #19221 .

`Does not need to be in the release notes now.`

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AntonPiccardoSelg** [comment]
@martyngigg  The system test is fixed. The export for the `MaskWorkspaceProperty` to Python had been missing.

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py [hunk 1]
```diff
@@ -451,17 +464,19 @@ def _move_monitor_4(workspace, move_info):
             detector_position = lab_detector_component.getPos()
             z_position_detector = detector_position.getZ()
 
-            monitor_4_offset = move_info.monitor_4_offset / 1000.
+            monitor_4_offset = move_info.monitor_4_offset
```
Comment (martyngigg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: Was the `/1000` was removed intentionally?

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py [hunk 2]
```diff
@@ -538,7 +555,9 @@ class SANSMoveLARMOROldStyle(SANSMove):
     def __init__(self):
         super(SANSMoveLARMOROldStyle, self).__init__()
 
-    def do_move_initial(self, move_info, workspace, coordinates, component):
+    def do_move_initial(self, move_info, workspace, coordinates, component, is_transmission_workspace):
+        _is_transmission_workspace = is_transmission_workspace  # noqa
```
Comment (martyngigg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: What does this line do?

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py [hunk 3]
```diff
@@ -597,7 +616,9 @@ def _rotate_around_y_axis(move_info, workspace, angle, component, bench_rotation
                      CanonicalCoordinates.Z: 0.0}
         rotate_component(workspace, total_angle, direction, detector_name)
 
-    def do_move_initial(self, move_info, workspace, coordinates, component):
+    def do_move_initial(self, move_info, workspace, coordinates, component, is_transmission_workspace):
+        _is_transmission_workspace = is_transmission_workspace  # noqa
```
Comment (martyngigg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: Same question as above.

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py [hunk 4]
```diff
@@ -611,7 +632,7 @@ def do_move_initial(self, move_info, workspace, coordinates, component):
                                     DetectorType.to_string(DetectorType.LAB))
 
         # Shift the low-angle bank detector in the x direction
-        angle = -coordinates[0]
+        angle = coordinates[0]
```
Comment (martyngigg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: Was the minus sign removed intentionally?

**Review by martyngigg — COMMENTED** [review]
Initial batch of comments.

File: scripts/test/SANS/user_file/user_file_state_director_test.py [hunk 5]
```diff
@@ -220,3 +217,6 @@ def test_stat_can_be_crated_from_valid_user_file_and_later_on_reset(self):
 
 if __name__ == "__main__":
     unittest.main()
+
```
Comment (martyngigg) [inline scripts/test/SANS/user_file/user_file_state_director_test.py]: It looks like we've duplicated this block.

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py (hunk 1 above)
Comment (AntonPiccardoSelg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: Yes, we convert mm to m already in the state builder, hence this was a mistake to convert it again.

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py (hunk 2 above)
Comment (AntonPiccardoSelg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: The methods is defined on an abstract base class and that particular field is used at the moment only by `LOQ` in `SANSMoveLOQ`.

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py (hunk 3 above)
Comment (AntonPiccardoSelg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: See above

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py (hunk 4 above)
Comment (AntonPiccardoSelg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: Yes, comparing it again with the old script showed that the the minus sign was wrong here.

File: scripts/test/SANS/user_file/user_file_state_director_test.py (hunk 5 above)
Comment (AntonPiccardoSelg) [inline scripts/test/SANS/user_file/user_file_state_director_test.py]: hmm, I am wondering if this has happened during a merge. I have had some issues with merges going silently wrong. I will remove it. Thanks for spotting this

File: scripts/SANS/sans/algorithm_detail/move_workspaces.py (hunk 2 above)
Comment (martyngigg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: Does flake8 complain if it's not used?

Comment (AntonPiccardoSelg) [inline scripts/SANS/sans/algorithm_detail/move_workspaces.py]: Yes it complains. I had to turn it off via `# noqa`

**Review by martyngigg — APPROVED** [review]
I think this is good to go now.

LGTM :shipit:

**Pull request merged**